presto-main/src/main/java/com/facebook/presto/operator/aggregation/estimatendv/SingleNDVEstimatorState.java

```diff
@@ -13,11 +13,78 @@
  */
 package com.facebook.presto.operator.aggregation.estimatendv;
 
+import com.facebook.presto.common.array.LongBigArray;
+import com.facebook.presto.common.type.Type;
+import com.facebook.presto.operator.aggregation.histogram.SingleHistogramState;
+import com.facebook.presto.operator.aggregation.histogram.SingleTypedHistogram;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.stream.LongStream;
+
 public class SingleNDVEstimatorState
-        extends NDVEstimatorState
+        extends SingleHistogramState
+        implements NDVEstimatorState
 {
-    public SingleNDVEstimatorState()
+    public SingleNDVEstimatorState(Type keyType, int expectedEntriesCount)
+    {
+        super(keyType, expectedEntriesCount);
+    }
+
+    private static Map<Long, Long> computeFreqDict(SingleTypedHistogram histogram)
+    {
+        LongBigArray counts = histogram.getCounts();
+        Map<Long, Long> freqDict = new HashMap<>();
+        int ndv = histogram.getDistinctValueCount();
+        for (int i = 0; i < ndv; i++) {
+            Long count = counts.get(i);
+            freqDict.put(count, freqDict.getOrDefault(count, 0L) + 1);
+        }
+        return freqDict;
+    }
+
+    private long computeBirthdayProblemProbability(long d)
+    {
+        long i = d;
+        long lowerBound;
+        while (true) {
+            lowerBound = d + i;
+            long finalLowerBound = lowerBound;
+            double multiplied = LongStream.range(0, d).mapToDouble(j -> (double) (finalLowerBound - j) / finalLowerBound).reduce(1, (x, y) -> x * y);
+            if (multiplied > LOWER_BOUND_PROBABILITY) {
+                break;
+            }
+            i += 1;
+        }
+        return lowerBound;
+    }
+
+    /**
+     * Use Chao estimator to estimate NDV
+     *
+     * @return Estimate of the NDV in the specific column in the original table
+     */
+    public long estimate()
     {
-        super();
+        Map<Long, Long> freqDict = computeFreqDict((SingleTypedHistogram) this.get());
```

Comment:
not sure this cast is always safe. I would have an `instanceof` check before and throw and exception if it is not an instance of `SingleTypedHistogram`

Reply:
This should be safe cause `this.get()` returns a `SingleTypedHistogram` value. The method signature of `this.get()` is TypedHistogram cause it is overridden from HistogramState